lib/ruby_lsp/event_emitter.rb

```diff
@@ -0,0 +1,29 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  class EventEmitter < SyntaxTree::Visitor
+    extend T::Sig
+
+    sig { params(listeners: Listener).void }
+    def initialize(*listeners)
+      @listeners = listeners
+
+      super()
+    end
+
+    # Emit events for a specific node. This is similar to the regular `visit` method, but avoids going deeper into the
+    # tree for performance
+    sig { params(node: T.nilable(SyntaxTree::Node)).void }
+    def emit_for_position(node)
+      case node
```

Comment:
I wonder if we should also trigger a `on_node` so the listener can subscribe to the one they want depending on the level of precision they need.

For example a listener that does the same thing for `Command` and `CallNode` could just listen to `on_node` and replicate the switch statement you have here rather than duplicating code across `on_command` and `on_call`.

This may be something we can add later if needed.